File: scripts/cbd_party_language_count.R
#!/usr/bin/env Rscript
# cbd_party_language_count.R
# Count languages per CBD Party country.
#
# Reads cbd_party_languages.csv and cbd_parties.csv, counts unique
# language-level glottocodes per country, and writes the result to
# data/cbd_party_language_count.csv.
#
# Output columns: country_name, iso_a2, iso_a3, language_count
#
# This produces the same output as the Python version
# (scripts/cbd_party_language_count.py).

library(tidyverse)
library(stringi)

# ── Paths ──────────────────────────────────────────────────────────────────────

project_root <- here::here()

cbd_languages_file <- file.path(project_root, "data", "cbd_party_languages.csv")
cbd_parties_file   <- file.path(project_root, "data", "cbd_parties.csv")
output_file        <- file.path(project_root, "data", "cbd_party_language_count.csv")

# ── Load party lookup (name + ISO_A3 from cbd_parties.csv) ─────────────────────

ISO2_FIXES <- c(
  "C\u00f4te d'Ivoire" = "CI",
  "T\u00fcrkiye"       = "TR",
  "Namibia"             = "NA"
)

cbd_raw <- read_csv(cbd_parties_file, show_col_types = FALSE, na = "")

party_lookup <- cbd_raw %>%
  mutate(
    Country = str_trim(Country),
    `ISO_A2` = str_trim(`ISO_A2`),
    `ISO_A3` = str_trim(`ISO_A3`),
    Party = str_trim(Party),
    Country_nfc = stri_trans_nfc(Country),
    `ISO_A2` = case_when(
      (`ISO_A2` == "" | is.na(`ISO_A2`)) & Country_nfc == "C\u00f4te d'Ivoire" ~ "CI",
      (`ISO_A2` == "" | is.na(`ISO_A2`)) & Country_nfc == "T\u00fcrkiye"         ~ "TR",
      (`ISO_A2` == "" | is.na(`ISO_A2`)) & Country == "Namibia"                  ~ "NA",
      TRUE                                                                         ~ `ISO_A2`
    )
  ) %>%
  filter(Party != "" & !is.na(Party)) %>%
  filter(!is.na(`ISO_A2`) & `ISO_A2` != "") %>%
  distinct(`ISO_A2`, .keep_all = TRUE) %>%
  select(country_name = Country, iso_a2 = `ISO_A2`, iso_a3 = `ISO_A3`)

cat(sprintf("Loaded %d CBD Party countries\n", nrow(party_lookup)))

# ── Count languages per country ────────────────────────────────────────────────

cbd_languages <- read_csv(cbd_languages_file, show_col_types = FALSE, na = "") %>%
  filter(`core.level` == "language") %>%
  select(glottocode, `core.countries`) %>%
  # Explode pipe-separated country codes
  mutate(country_codes = str_split(str_trim(`core.countries`), "\\s*\\|\\s*")) %>%
  unnest(country_codes) %>%
  filter(country_codes != "") %>%
  # Count unique glottocodes per country
  group_by(iso_a2 = country_codes) %>%
  summarise(language_count = n_distinct(glottocode), .groups = "drop")

# ── Join with party lookup and sort ─────────────────────────────────────────────

result <- party_lookup %>%
  inner_join(cbd_languages, by = "iso_a2") %>%
  arrange(desc(language_count), iso_a2)

cat(sprintf("Wrote %d countries\n", nrow(result)))

# ── Write output ───────────────────────────────────────────────────────────────

write_csv(result, output_file)
cat(sprintf("Output written to %s\n", output_file))

File: scripts/cbd_party_language_count.py
#!/usr/bin/env python3
"""
Count languages per CBD Party country.

Reads cbd_party_languages.csv and cbd_parties.csv, counts unique language-level
glottocodes per country, and writes the result to data/cbd_party_language_count.csv.

Output columns: country_name, iso_a2, iso_a3, language_count

This produces the same output as the R version (scripts/cbd_party_language_count.R).
"""

import csv
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent
CBD_LANGUAGES_FILE = PROJECT_ROOT / "data" / "cbd_party_languages.csv"
CBD_PARTIES_FILE = PROJECT_ROOT / "data" / "cbd_parties.csv"
OUTPUT_FILE = PROJECT_ROOT / "data" / "cbd_party_language_count.csv"


def load_party_lookup(filepath):
    """Load a dict mapping ISO_A2 -> (country_name, ISO_A3) from cbd_parties.csv.

    Handles Unicode combining characters, missing ISO_A2 codes (CI, TR, NA),
    and strips whitespace from all fields. Deduplicates by ISO_A2.
    """
    import unicodedata

    ISO2_FIXES = {
        unicodedata.normalize("NFC", "Côte d'Ivoire"): "CI",
        unicodedata.normalize("NFC", "Türkiye"): "TR",
        "Namibia": "NA",
    }

    lookup = {}
    with open(filepath, "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            party_date = (row.get("Party") or "").strip()
            if not party_date:
                continue

            code = (row.get("ISO_A2") or "").strip()
            country = (row.get("Country") or "").strip()
            iso_a3 = (row.get("ISO_A3") or "").strip()

            if not code:
                country_nfc = unicodedata.normalize("NFC", country)
                if country_nfc in ISO2_FIXES:
                    code = ISO2_FIXES[country_nfc]

            if not code:
                continue

            if code not in lookup:
                lookup[code] = (country, iso_a3)

    return lookup


def count_languages_per_party(cbd_languages_path, party_lookup):
    """Count unique language-level glottocodes per CBD Party ISO_A2 code."""
    counts = {}
    with open(cbd_languages_path, "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            if row.get("core.level", "") != "language":
                continue
            raw = (row.get("core.countries") or "").strip()
            if not raw:
                continue
            codes = {c.strip() for c in raw.split("|") if c.strip()}
            for code in codes:
                if code in party_lookup:
                    counts[code] = counts.get(code, 0) + 1

    return counts


def main():
    party_lookup = load_party_lookup(CBD_PARTIES_FILE)
    counts = count_languages_per_party(CBD_LANGUAGES_FILE, party_lookup)

    # Build output rows sorted by language count descending
    rows = []
    for code, count in sorted(counts.items(), key=lambda x: (-x[1], x[0])):
        country_name, iso_a3 = party_lookup[code]
        rows.append({
            "country_name": country_name,
            "iso_a2": code,
            "iso_a3": iso_a3,
            "language_count": count,
        })

    with open(OUTPUT_FILE, "w", encoding="utf-8", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=["country_name", "iso_a2", "iso_a3", "language_count"])
        writer.writeheader()
        for row in rows:
            writer.writerow(row)

    print(f"Wrote {len(rows)} countries to {OUTPUT_FILE}")
    total_languages = sum(counts.values())
    print(f"Total language-country pairs: {total_languages}")


if __name__ == "__main__":
    main()
